### PR TITLE
PTExecutor: Add PTrix-based PT Executor Class

### PR DESCRIPTION
**Pull request by retrage**
## Type of PR

- [x] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [ ] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [ ] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Related Issue

See #12.

## Importance of PR

- Importance of the issue
    - [ ] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [x] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [ ] Large (requires several days to several weeks of review)
    - [x] Medium (requires several hours to a day of review)
    - [ ] Small (trivial changes, such as build error)

## PR Overview

https://github.com/fuzzuf/fuzzuf/issues/12#issuecomment-1031086928

## Concerns (Optional)

- [ ] Performance
- [ ] Source Code Quality

This PR adds PTrix-based PT Executor, but it does not contain tests due to the following reasons.
1. It is hard to test the PT executor because it requires:
  - PT control kernel module `ptmodule`. It cannot be tested in VMs or containers.
  - Proxy application `pt-proxy-fast`. This application is still flaky and not stable.
  - `patchelf` to PUT so that the PUT run as a fork server.
I tested this PR feature on my local environment, but it is not ready for push here.

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [ ] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [x] The PR title is a summary of the changes.
- [x] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [ ] CI is green or confirmed test run results.
- [ ] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [ ] The title of this PR summarizes the changes made by this PR properly.
- [ ] The target branch of this PR is as intended.
- [ ] The reviewer understands the issues in this PR.
- [ ] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

### CODE REVIEW

**Comment by K-atc** [comment]
@retrage I would like to do a design review first.
Please answer the following questions (I could not figure out them from the PR or Issue description section. Please be aware of the following in the next issue).

* (1) There seems to be no documentation attached in this PR. This is difficult for a third party to use, is this intentional?

* (2) What did you see that led you to believe this Executor was completed? Please provide evidence.
    * In most cases, it is sufficient to provide a log of the results of unit tests and CLI actions
    * If you don't have a policy, please feel free to contact me. If you don't have a policy, please feel free to contact consult me, and let's find where to land together.

* (3) Are there no unit tests? Can you check the completeness of the functionality with a visual code review?
    
* (4) Please clarify the dependencies: does this Executor work with the fuzzuf repository and [Requirements of fuzzuf](https://github.com/fuzzuf/fuzzuf/blob/master/TUTORIAL.md)?

File: check/config.h.in [hunk 1]
```diff
@@ -31,5 +31,7 @@
 #define FUZZUF_QEMU_EXECUTABLE "${QEMU_EXECUTABLE}"
 // NOTE: FUZZUF_CS_PROXY_EXECUTABLE is for testing purpose only. Do not use in the Fuzzuf core implementation.
 #define FUZZUF_CS_PROXY_EXECUTABLE "${CS_PROXY_EXECUTABLE}"
+// NOTE: FUZZUF_PT_PROXY_EXECUTABLE is for testing purpose only. Do not use in the Fuzzuf core implementation.
+#define FUZZUF_PT_PROXY_EXECUTABLE "${PT_PROXY_EXECUTABLE}"
```
Comment (K-atc) [inline check/config.h.in:35]: (No action required) More broken windows. I'd like to solve the problem of not being able to distribute fuzzuf as a package because of the runtime settings at the time of this build, but... I'll make a task in Notion.

File: cmake/FindPTProxy.cmake [hunk 2]
```diff
@@ -0,0 +1,40 @@
+#
+# fuzzuf
+# Copyright (C) 2021 Ricerca Security
```
Comment (K-atc) [inline cmake/FindPTProxy.cmake:3]: License notation exists. Yoshi!

File: executor/pt_executor.cpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2021 Ricerca Security
```
Comment (K-atc) [inline executor/pt_executor.cpp:3]: License notation exists. Yoshi!

File: executor/pt_executor.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+#include "fuzzuf/executor/pt_executor.hpp"
+
+#include "fuzzuf/logger/logger.hpp"
+
+// NOTE:
+//    - PTExecutor assume it can create a file at `path_to_write_input`.
+PTExecutor::PTExecutor(
+    const fs::path &proxy_path,
```
Comment (K-atc) [inline executor/pt_executor.cpp:25]: Designed the path of the application to be proxied. Good

File: include/fuzzuf/executor/proxy_executor.hpp [hunk 5]
```diff
@@ -171,7 +171,11 @@ class ProxyExecutor : public Executor {
     // InplaceMemoryFeedback made of GetStdErr before calling this function becomes invalid after Run()
     fuzzuf::executor::output_t MoveStdErr();
 
-private:    
+protected:
+    bool record_stdout_and_err;
+    bool has_shared_memories;
```
Comment (K-atc) [inline include/fuzzuf/executor/proxy_executor.hpp:176]: (nits) Isn't it customary to use the singular form of "memory" for storage devices?
The word "memories" means 思い出.

File: include/fuzzuf/executor/pt_executor.hpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2021 Ricerca Security
```
Comment (K-atc) [inline include/fuzzuf/executor/pt_executor.hpp:3]: License notation exists. Yoshi!

File: include/fuzzuf/executor/pt_executor.hpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    static constexpr u32 FAV_SHM_SIZE = (1U << 16);
+
+    const u32 path_shm_size;
+    const u32 fav_shm_size;
+
+    int path_shmid;
+    int fav_shmid;
+
+    u8 *path_trace_bits;
+    u8 *fav_trace_bits;
```
Comment (K-atc) [inline include/fuzzuf/executor/pt_executor.hpp:47]: (Curious question) What role do `path_shm*` and `fav_shm*` play in PTExecutor?

File: include/fuzzuf/executor/pt_executor.hpp [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+#pragma once
+
+#include "fuzzuf/executor/proxy_executor.hpp"
+
+// A class for fuzz executions with Intel PT
+class PTExecutor : public ProxyExecutor {
+public:
+    static constexpr const char* PATH_SHM_ENV_VAR = "__AFL_SHM_ID"; // TODO: Use different environment variable.
```
Comment (K-atc) [inline include/fuzzuf/executor/pt_executor.hpp:33]: (Curious question）Why `Use different environment variable` is required?

**Review by K-atc — COMMENTED** [review]
I performed code review. There are some questions, but may no need to change codes.

- `(nits)` is クソリプ and you can ignore it.
- [x] Evidence that proofs this Executor works in retrage's local environment

**Comment by retrage** [comment]
TL;DR
1. The dependency repository is not yet public and is not intended to be used by third parties at this time.
2. We'll post the results of local testing as evidence.
3. Due to various limitations, automated testing is difficult.
4. Because the dependency repository is not yet public.

I haven't intentionally added any tests to this PR. This is because it is difficult to test for the following reasons:
1. Intel PT and the `ptmodule` kernel module that controls it are required.
2. A proxy application, `pt-proxy-fast`, is needed.
3. We need to `patchelf` the PUT and run it as a fork server with a modified glibc.

As for 3., CoreSight implemented non-fork server mode, so we were able to avoid it for testing, but `pt-proxy-fast` is not supported, so it is a requirement. For this reason, a modified glibc is also required to have what we need for testing.
As mentioned above, the dependency repositories are not open to the public and are not intended to be used by third parties, so there is no documentation available.
These are currently managed in a private repository. We have not yet decided what to do with this repository, but we think it will be managed in a separate repository.
It is possible to put the test code that we have checked in our local environment, but at this point, it is still not possible to verify the code with automated tests, so we decided that it is not significant to include it in this PR. I think it is important that this PR is built with PT Executor and that there are no regressions.

File: include/fuzzuf/executor/proxy_executor.hpp (hunk 5 above)
Comment (retrage) [reply include/fuzzuf/executor/proxy_executor.hpp:176]: I'm not sure the word "memory" is countable or uncountable. I just followed existing code like "SetupSharedMemories".

**Comment by retrage** [comment]
Below log is evidence that this PT executor passed basic unit tests on my local machine.
```
$ sudo ./test/executor/pt_executor/test-pt-executor-run
Running 1 test case...
[*] output_file_path = /tmp/fuzzuf_test.xpWRox/output/result
cargv[0]: /home/akira/src/pt/AFLplusplus/pt-proxy-fast
Run: 
/home/akira/src/pt/AFLplusplus/pt-proxy-fast 
/home/akira/src/pt/AFLplusplus/pt-proxy-fast 
-- 
/home/akira/local/bin/tee.patched 
/tmp/fuzzuf_test.xpWRox/output/result 
(null) 


Exec Status 0 (pid 0)


*** No errors detected
```

The test code patch is here:
```diff
diff --git a/test/executor/CMakeLists.txt b/test/executor/CMakeLists.txt
index 446a57c..a01ace1 100644
--- a/test/executor/CMakeLists.txt
+++ b/test/executor/CMakeLists.txt
@@ -50,6 +50,7 @@ subdirs(
   proxy_executor
   qemu_executor
   coresight_executor
+  pt_executor
 )
 
 if(PIN_FOUND)
diff --git a/test/executor/pt_executor/CMakeLists.txt b/test/executor/pt_executor/CMakeLists.txt
new file mode 100644
index 0000000..79056d5
--- /dev/null
+++ b/test/executor/pt_executor/CMakeLists.txt
@@ -0,0 +1,31 @@
+if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "x86_64")
+  add_executable( test-pt-executor-run run.cpp )
+  target_link_libraries(
+    test-pt-executor-run
+    test-common
+    fuzzuf
+    ${FUZZUF_LIBRARIES}
+    Boost::unit_test_framework
+  )
+  target_include_directories(
+    test-pt-executor-run
+    PRIVATE
+    ${FUZZUF_INCLUDE_DIRS}
+    ${CMAKE_SOURCE_DIR}/test/common
+  )
+  set_target_properties(
+    test-pt-executor-run
+    PROPERTIES COMPILE_FLAGS "${ADDITIONAL_COMPILE_FLAGS_STR}"
+  )
+  set_target_properties(
+    test-pt-executor-run
+    PROPERTIES LINK_FLAGS "${ADDITIONAL_COMPILE_FLAGS_STR}"
+  )
+endif()
+
+if(PT_TRACE_FOUND)
+  add_test(
+    NAME "pt_executor.run"
+    COMMAND sudo -E $<TARGET_FILE:test-pt-executor-run>
+  )
+endif()
diff --git a/test/executor/pt_executor/run.cpp b/test/executor/pt_executor/run.cpp
new file mode 100644
index 0000000..5834c79
--- /dev/null
+++ b/test/executor/pt_executor/run.cpp
@@ -0,0 +1,99 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2021 Ricerca Security
+ * 
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+#define BOOST_TEST_MODULE pt_executor.run
+#define BOOST_TEST_DYN_LINK
+
+#include <boost/scope_exit.hpp>
+#include <boost/test/unit_test.hpp>
+#include <iostream>
+
+#include "fuzzuf/algorithms/afl/afl_option.hpp"
+#include "fuzzuf/exceptions.hpp"
+#include "fuzzuf/executor/pt_executor.hpp"
+#include "fuzzuf/feedback/inplace_memory_feedback.hpp"
+#include "fuzzuf/feedback/put_exit_reason_type.hpp"
+#include "fuzzuf/utils/common.hpp"
+#include "fuzzuf/utils/filesystem.hpp"
+#include "fuzzuf/utils/which.hpp"
+
+// PTExecutor::Run() の正常系テスト
+BOOST_AUTO_TEST_CASE(PTExecutorRun) {
+  // Setup root directory
+  std::string root_dir_template("/tmp/fuzzuf_test.XXXXXX");
+  const auto raw_dirname = mkdtemp(root_dir_template.data());
+  if (!raw_dirname)
+    throw -1;
+  BOOST_CHECK(raw_dirname != nullptr);
+  auto root_dir = fs::path(raw_dirname);
+  BOOST_SCOPE_EXIT(&root_dir) { fs::remove_all(root_dir); }
+  BOOST_SCOPE_EXIT_END
+
+  // Create input/output dirctory
+  auto input_dir = root_dir / "input";
+  auto output_dir = root_dir / "output";
+  BOOST_CHECK_EQUAL(fs::create_directory(input_dir), true);
+  BOOST_CHECK_EQUAL(fs::create_directory(output_dir), true);
+
+  // Setup output file path
+  auto output_file_path = output_dir / "result";
+  std::cout << "[*] output_file_path = " + output_file_path.native()
+            << std::endl;
+
+  auto path_to_write_seed = output_dir / "cur_input";
+  PTExecutor executor(fs::path(FUZZUF_PT_PROXY_EXECUTABLE),
+                             {"/home/akira/local/bin/tee.patched", output_file_path.native()},
+                             1000,
+                             10000, true, path_to_write_seed,
+                             PTExecutor::CPUID_DO_NOT_BIND,
+                             true /* record_stdout_and_err */
+  );
+  BOOST_CHECK_EQUAL(executor.stdin_mode, true);
+
+  // Invoke PTExecutor::Run()
+  std::string input("abcd");
+  executor.Run(reinterpret_cast<const u8 *>(input.c_str()), input.size());
+
+  // Check normality
+  // (1) 正常実行されたこと → feedbackのexit_reason が
+  // PUTExitReasonType::FAULT_NONE であることを確認する
+  BOOST_CHECK_EQUAL(executor.GetExitStatusFeedback().exit_reason,
+                    PUTExitReasonType::FAULT_NONE);
+
+  // (2) 標準入力によってファズが受け渡されたこと →
+  // 標準入力と同じ内容がファイルに保存されたことを確認する
+  auto stdout_buffer = executor.MoveStdOut();
+  std::vector<std::uint8_t> expected_stdout{'a', 'b', 'c', 'd'};
+  BOOST_CHECK_EQUAL_COLLECTIONS(stdout_buffer.begin(), stdout_buffer.end(),
+                                expected_stdout.begin(), expected_stdout.end());
+  // (3) Check if the executor wrote the shared memory feedback.
+  BOOST_CHECK(executor.GetPathFeedback().CountNonZeroBytes() > 0);
+  BOOST_CHECK(executor.GetFavFeedback().CountNonZeroBytes() > 0);
+
+  // Ignore stderr output for PTExecutor because the tracer often outputs
+  // debug messages with memory map info.
+#if 0
+  auto stderr_buffer = executor.MoveStdErr();
+  std::vector< std::uint8_t > expected_stderr{};
+  BOOST_CHECK_EQUAL_COLLECTIONS(
+    stderr_buffer.begin(),
+    stderr_buffer.end(),
+    expected_stderr.begin(),
+    expected_stderr.end()
+  );
+#endif
+}
```

**Comment by retrage** [comment]
Below log is evidence that this PT executor passed basic unit tests on my local machine. This test runs `patchelf`ed tee command and checks the feedback.
```
$ sudo ./test/executor/pt_executor/test-pt-executor-run
Running 1 test case...
[*] output_file_path = /tmp/fuzzuf_test.xpWRox/output/result
cargv[0]: /home/akira/src/pt/AFLplusplus/pt-proxy-fast
Run: 
/home/akira/src/pt/AFLplusplus/pt-proxy-fast 
/home/akira/src/pt/AFLplusplus/pt-proxy-fast 
-- 
/home/akira/local/bin/tee.patched 
/tmp/fuzzuf_test.xpWRox/output/result 
(null) 


Exec Status 0 (pid 0)


*** No errors detected
```

The test code patch is here:
```diff
diff --git a/test/executor/CMakeLists.txt b/test/executor/CMakeLists.txt
index 446a57c..a01ace1 100644
--- a/test/executor/CMakeLists.txt
+++ b/test/executor/CMakeLists.txt
@@ -50,6 +50,7 @@ subdirs(
   proxy_executor
   qemu_executor
   coresight_executor
+  pt_executor
 )
 
 if(PIN_FOUND)
diff --git a/test/executor/pt_executor/CMakeLists.txt b/test/executor/pt_executor/CMakeLists.txt
new file mode 100644
index 0000000..79056d5
--- /dev/null
+++ b/test/executor/pt_executor/CMakeLists.txt
@@ -0,0 +1,31 @@
+if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "x86_64")
+  add_executable( test-pt-executor-run run.cpp )
+  target_link_libraries(
+    test-pt-executor-run
+    test-common
+    fuzzuf
+    ${FUZZUF_LIBRARIES}
+    Boost::unit_test_framework
+  )
+  target_include_directories(
+    test-pt-executor-run
+    PRIVATE
+    ${FUZZUF_INCLUDE_DIRS}
+    ${CMAKE_SOURCE_DIR}/test/common
+  )
+  set_target_properties(
+    test-pt-executor-run
+    PROPERTIES COMPILE_FLAGS "${ADDITIONAL_COMPILE_FLAGS_STR}"
+  )
+  set_target_properties(
+    test-pt-executor-run
+    PROPERTIES LINK_FLAGS "${ADDITIONAL_COMPILE_FLAGS_STR}"
+  )
+endif()
+
+if(PT_TRACE_FOUND)
+  add_test(
+    NAME "pt_executor.run"
+    COMMAND sudo -E $<TARGET_FILE:test-pt-executor-run>
+  )
+endif()
diff --git a/test/executor/pt_executor/run.cpp b/test/executor/pt_executor/run.cpp
new file mode 100644
index 0000000..5834c79
--- /dev/null
+++ b/test/executor/pt_executor/run.cpp
@@ -0,0 +1,99 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2021 Ricerca Security
+ * 
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+#define BOOST_TEST_MODULE pt_executor.run
+#define BOOST_TEST_DYN_LINK
+
+#include <boost/scope_exit.hpp>
+#include <boost/test/unit_test.hpp>
+#include <iostream>
+
+#include "fuzzuf/algorithms/afl/afl_option.hpp"
+#include "fuzzuf/exceptions.hpp"
+#include "fuzzuf/executor/pt_executor.hpp"
+#include "fuzzuf/feedback/inplace_memory_feedback.hpp"
+#include "fuzzuf/feedback/put_exit_reason_type.hpp"
+#include "fuzzuf/utils/common.hpp"
+#include "fuzzuf/utils/filesystem.hpp"
+#include "fuzzuf/utils/which.hpp"
+
+// PTExecutor::Run() の正常系テスト
+BOOST_AUTO_TEST_CASE(PTExecutorRun) {
+  // Setup root directory
+  std::string root_dir_template("/tmp/fuzzuf_test.XXXXXX");
+  const auto raw_dirname = mkdtemp(root_dir_template.data());
+  if (!raw_dirname)
+    throw -1;
+  BOOST_CHECK(raw_dirname != nullptr);
+  auto root_dir = fs::path(raw_dirname);
+  BOOST_SCOPE_EXIT(&root_dir) { fs::remove_all(root_dir); }
+  BOOST_SCOPE_EXIT_END
+
+  // Create input/output dirctory
+  auto input_dir = root_dir / "input";
+  auto output_dir = root_dir / "output";
+  BOOST_CHECK_EQUAL(fs::create_directory(input_dir), true);
+  BOOST_CHECK_EQUAL(fs::create_directory(output_dir), true);
+
+  // Setup output file path
+  auto output_file_path = output_dir / "result";
+  std::cout << "[*] output_file_path = " + output_file_path.native()
+            << std::endl;
+
+  auto path_to_write_seed = output_dir / "cur_input";
+  PTExecutor executor(fs::path(FUZZUF_PT_PROXY_EXECUTABLE),
+                             {"/home/akira/local/bin/tee.patched", output_file_path.native()},
+                             1000,
+                             10000, true, path_to_write_seed,
+                             PTExecutor::CPUID_DO_NOT_BIND,
+                             true /* record_stdout_and_err */
+  );
+  BOOST_CHECK_EQUAL(executor.stdin_mode, true);
+
+  // Invoke PTExecutor::Run()
+  std::string input("abcd");
+  executor.Run(reinterpret_cast<const u8 *>(input.c_str()), input.size());
+
+  // Check normality
+  // (1) 正常実行されたこと → feedbackのexit_reason が
+  // PUTExitReasonType::FAULT_NONE であることを確認する
+  BOOST_CHECK_EQUAL(executor.GetExitStatusFeedback().exit_reason,
+                    PUTExitReasonType::FAULT_NONE);
+
+  // (2) 標準入力によってファズが受け渡されたこと →
+  // 標準入力と同じ内容がファイルに保存されたことを確認する
+  auto stdout_buffer = executor.MoveStdOut();
+  std::vector<std::uint8_t> expected_stdout{'a', 'b', 'c', 'd'};
+  BOOST_CHECK_EQUAL_COLLECTIONS(stdout_buffer.begin(), stdout_buffer.end(),
+                                expected_stdout.begin(), expected_stdout.end());
+  // (3) Check if the executor wrote the shared memory feedback.
+  BOOST_CHECK(executor.GetPathFeedback().CountNonZeroBytes() > 0);
+  BOOST_CHECK(executor.GetFavFeedback().CountNonZeroBytes() > 0);
+
+  // Ignore stderr output for PTExecutor because the tracer often outputs
+  // debug messages with memory map info.
+#if 0
+  auto stderr_buffer = executor.MoveStdErr();
+  std::vector< std::uint8_t > expected_stderr{};
+  BOOST_CHECK_EQUAL_COLLECTIONS(
+    stderr_buffer.begin(),
+    stderr_buffer.end(),
+    expected_stderr.begin(),
+    expected_stderr.end()
+  );
+#endif
+}
```

File: include/fuzzuf/executor/proxy_executor.hpp (hunk 5 above)
Comment (K-atc) [reply include/fuzzuf/executor/proxy_executor.hpp:176]: 👌

**Comment by K-atc** [comment]
Executor seems to be working.

File: include/fuzzuf/executor/pt_executor.hpp (hunk 7 above)
Comment (retrage) [reply include/fuzzuf/executor/pt_executor.hpp:47]: Comments added with 82a84cf answer this question.

File: include/fuzzuf/executor/pt_executor.hpp (hunk 8 above)
Comment (retrage) [reply include/fuzzuf/executor/pt_executor.hpp:33]: Please see comments in [82a84cf](https://github.com/fuzzuf/fuzzuf/pull/13/commits/82a84cf2a2117b9a14fa1d1ac2e3180aaeeee52b).

Comment (K-atc) [reply include/fuzzuf/executor/pt_executor.hpp:33]: Perfect! Thanks!

File: include/fuzzuf/executor/pt_executor.hpp (hunk 7 above)
Comment (K-atc) [reply include/fuzzuf/executor/pt_executor.hpp:47]: Perfect! Thanks!

**Review by K-atc — APPROVED** [review]
お疲れさまでした～

> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [x] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [x] If minor refactorings are not mentioned in the PR, I understand the intent.
- [x] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [x] The reviewer understands how this PR addresses the issue and the specific changes.
- [x] This PR uses the best possible issue resolution that the reviewer can think of.
- [x] This PR is ready to be merged.
